### PR TITLE
Migrated the Gitter and MailChimp authentication providers to ASP.NET 2.0.

### PR DESCRIPTION
**Pull request by haydenhancock**
Migrated the Gitter and MailChimp authentication providers to ASP.NET 2.0.

### CODE REVIEW

**Comment by haydenhancock** [comment]
@PinpointTownes How do I fix this?

**Comment by kinosang** [comment]
Your changes has conflicting files to the upstream (it may be caused by other pull requests that already merged or new commits to the upstream).

Pull from or rebase to the upstream and resolve the conflicting manually, then you should `force push` it.

If it's difficult to be done, you can simply backup your changes, create (clone) a new repo from upstream, create new branch, restore your backup, `force push` it to your online repo (not the upstream).

File: src/AspNet.Security.OAuth.Gitter/AspNet.Security.OAuth.Gitter.csproj [hunk 1]
```diff
@@ -3,21 +3,25 @@
   <Import Project="..\..\build\packages.props" />
 
   <PropertyGroup>
-    <TargetFrameworks>net451;netstandard1.3</TargetFrameworks>
+    <TargetFrameworks>netstandard2.0</TargetFrameworks>
```
Comment (kevinchalet) [inline src/AspNet.Security.OAuth.Gitter/AspNet.Security.OAuth.Gitter.csproj:6]: `<TargetFramework>netstandard2.0</TargetFramework>`

File: samples/Mvc.Client/Properties/launchSettings.json [hunk 2]
```diff
@@ -3,8 +3,8 @@
     "windowsAuthentication": false,
     "anonymousAuthentication": true,
     "iisExpress": {
-      "applicationUrl": "https://localhost:44318/",
-      "sslPort": 44318
+      "applicationUrl": "http://127.0.0.1:44318/",
```
Comment (kevinchalet) [inline samples/Mvc.Client/Properties/launchSettings.json:6]: Revert, please.

File: src/AspNet.Security.OAuth.Gitter/GitterAuthenticationDefaults.cs [hunk 3]
```diff
@@ -46,6 +47,6 @@ public static class GitterAuthenticationDefaults
         /// <summary>
         /// Default value for <see cref="OAuthOptions.UserInformationEndpoint"/>.
         /// </summary>
-        public const string UserInformationEndpoint = "https://api.gitter.im/v1/user";
+        public const string UserInformationEndpoint = "https://api.gitter.im/v1/user/me";
```
Comment (kevinchalet) [inline src/AspNet.Security.OAuth.Gitter/GitterAuthenticationDefaults.cs:50]: Functional change. Why is it needed?

File: src/AspNet.Security.OAuth.Gitter/GitterAuthenticationExtensions.cs [hunk 4]
```diff
@@ -4,10 +4,10 @@
  * for more information concerning the license and the contributors participating to this project.
  */
 
-using System;
 using AspNet.Security.OAuth.Gitter;
-using JetBrains.Annotations;
-using Microsoft.Extensions.Options;
+using Microsoft.AspNetCore.Authentication;
+using Microsoft.Extensions.DependencyInjection;
+using System;
```
Comment (kevinchalet) [inline src/AspNet.Security.OAuth.Gitter/GitterAuthenticationExtensions.cs:10]: `System` using always go first.

File: src/AspNet.Security.OAuth.Gitter/GitterAuthenticationExtensions.cs [hunk 5]
```diff
@@ -17,54 +17,60 @@ namespace Microsoft.AspNetCore.Builder
...
+        /// <see cref="AuthenticationBuilder"/>, which enables Gitter authentication capabilities.
         /// </summary>
-        /// <param name="app">The <see cref="IApplicationBuilder"/> to add the middleware to.</param>
-        /// <param name="options">A <see cref="GitterAuthenticationOptions"/> that specifies options for the middleware.</param>
+        /// <param name="builder">The authentication builder.</param>
         /// <returns>A reference to this instance after the operation has completed.</returns>
-        public static IApplicationBuilder UseGitterAuthentication(
-            [NotNull] this IApplicationBuilder app,
-            [NotNull] GitterAuthenticationOptions options)
+        public static AuthenticationBuilder AddGitter(this AuthenticationBuilder builder)
```
Comment (kevinchalet) [inline src/AspNet.Security.OAuth.Gitter/GitterAuthenticationExtensions.cs:25]: Why did you remove the JetBrains annotations?

File: src/AspNet.Security.OAuth.Gitter/GitterAuthenticationHandler.cs [hunk 6]
```diff
@@ -4,33 +4,36 @@
...
+using System.Threading.Tasks;
 
 namespace AspNet.Security.OAuth.Gitter
 {
     public class GitterAuthenticationHandler : OAuthHandler<GitterAuthenticationOptions>
     {
-        public GitterAuthenticationHandler([NotNull] HttpClient client)
-            : base(client)
+        public GitterAuthenticationHandler(
+            IOptionsMonitor<GitterAuthenticationOptions> options,
```
Comment (kevinchalet) [inline src/AspNet.Security.OAuth.Gitter/GitterAuthenticationHandler.cs:23]: Same remark. Please bring back the annotations.

File: src/AspNet.Security.OAuth.Gitter/AspNet.Security.OAuth.Gitter.csproj [hunk 7]
```diff
@@ -3,21 +3,25 @@
...
   </PropertyGroup>
 
   <PropertyGroup>
     <Description>ASP.NET Core security middleware enabling Gitter authentication.</Description>
-    <Authors>Maxime Roussin-Bélanger</Authors>
+    <Authors>Hayden Hancock</Authors>
```
Comment (kevinchalet) [inline src/AspNet.Security.OAuth.Gitter/AspNet.Security.OAuth.Gitter.csproj:11]: Please don't remove the name of the initial contributor. You can add yours while preserving the initial one'. E.g:

<Authors>Maxime Roussin-Bélanger;Hayden Hancock</Authors>

File: src/AspNet.Security.OAuth.Gitter/AspNet.Security.OAuth.Gitter.csproj [hunk 8]
```diff
@@ -3,21 +3,25 @@
...
     <PackageTags>aspnetcore;authentication;gitter;oauth;security</PackageTags>
+    <Company>Hayden Hancock</Company>
```
Comment (kevinchalet) [inline src/AspNet.Security.OAuth.Gitter/AspNet.Security.OAuth.Gitter.csproj:13]: Nope, remove.

File: src/AspNet.Security.OAuth.Gitter/AspNet.Security.OAuth.Gitter.csproj [hunk 9]
```diff
@@ -3,21 +3,25 @@
...
   </PropertyGroup>
 
   <ItemGroup>
     <Compile Include="..\..\shared\AspNet.Security.OAuth.Extensions\*.cs" />
   </ItemGroup>
 
   <ItemGroup>
-    <PackageReference Include="JetBrains.Annotations" Version="$(JetBrainsVersion)" PrivateAssets="All" />
+    <Compile Remove="..\..\shared\AspNet.Security.OAuth.Extensions\ClaimsExtensions.cs" />
```
Comment (kevinchalet) [inline src/AspNet.Security.OAuth.Gitter/AspNet.Security.OAuth.Gitter.csproj:21]: This doesn't look right at all. Take a look at the other providers and make that consistent, please.

**Review by kevinchalet — CHANGES_REQUESTED** [review]

**Comment by kevinchalet** [comment]
> @PinpointTownes How do I fix this?

Yep, rebasing as suggested by @kinosang is the right thing to do.

I added a few remarks about inconsistent things. Please address them when you have a minute :+1:

File: samples/Mvc.Client/Properties/launchSettings.json (hunk 2 above)
Comment (haydenhancock) [reply samples/Mvc.Client/Properties/launchSettings.json:6]: FYI, MailChimp doesn't allow you to use localhost.

File: src/AspNet.Security.OAuth.Gitter/GitterAuthenticationExtensions.cs (hunk 4 above)
Comment (haydenhancock) [reply src/AspNet.Security.OAuth.Gitter/GitterAuthenticationExtensions.cs:10]: Interesting. Visual Studio 2017 puts them in alphabetically order.

**Comment by haydenhancock** [comment]
I am trying to figure out how to rebase properly. I rebased with the experimental branch (from my migrated-providers branch) on my fork and it did prompt me to resolve some conflicts but now I am getting the following which doesn't look right:

image

Any guidance?

**Comment by kevinchalet** [comment]
Consider using the command line:

```git
git fetch aspnet-contrib
git checkout migrate-providers
git rebase aspnet-contrib/experimental
// Fix conflicts if necessary (call git add . and git rebase --continue when issues are fixed)
git push -f
```

**Comment by kevinchalet** [comment]
Rebased, fixed and merged. Thanks for your contribution :+1:
